### ISSUE TITLE
Your code currently operates one image at a time, is it possible to adjust the batch_size operation? I look forward to hearing from you!

### ISSUE THREAD

**Issue by tanlingp** [WindVChen/DiffAttack#6]
Your code currently operates one image at a time, is it possible to adjust the batch_size operation? I look forward to hearing from you!

**Comment by WindVChen** [comment]
Hi @tanlingp ,

In theory, we can of course generate adversarial images in batches. However, since this method currently requires backpropagation through multiple diffusion steps, many intermediate calculation graphs are stored, so even if only a single image is processed, the memory usage is approximately 16GB. Thus with only my current equipment (a single Rtx3090), I'm afraid I am not able to test turning the code into batches.

However, if you have multiple GPUs or more memory, the most straightforward way I recommend is to use `multiprocessing` to process batches of images. It's easy and doesn't require modifying the core code.

Hope this helps.

**Comment by yujinhan98** [comment]
Hi @WindVChen,

This [article](https://neurips.cc/virtual/2023/poster/71718) proposes the Segment-wise Forwarding-Backwarding algorithm to address the memory issue in computational graphs, aiming to improve the efficiency of DiffAttack.

**Comment by WindVChen** [comment]
Hi @yujinhan98,

Thank you for sharing the article. I appreciate your insight in addressing the memory issue with Segment-wise Forwarding-Backwarding. However, it seems that this approach might prolong the optimization time, which might still fail to speed up the generation of the adversarial examples.

While I haven't delved into all the intricacies of the article, Segment-wise Forwarding-Backwarding seems akin to the commonly used [gradient checkpointing trick](https://github.com/cybertronai/gradient-checkpointing). This method involves recalculating computational graphs during the backpropagation phase, leading to increased processing time, contingent on the number of checkpoints set.

@tanlingp's intention, I think, is to expedite the generation of adversarial examples by leveraging batch processing. Given the longer processing time per input associated with Segment-wise Forwarding-Backwarding (despite its reduced memory usage), it may not significantly enhance the speed, which is the primary concern here.

Nonetheless, I appreciate your contribution. Many thanks for sharing your insights.

Best regards.
